### PR TITLE
Dtype Analysis for Unary and Binary ops with Metatensors

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67786
* #67575
* #67500
* #66973
* #66948
* #66899
* __->__ #66898

Differential Revision: [D32175961](https://our.internmc.facebook.com/intern/diff/D32175961)

### CODE REVIEW

**Review by Krovatkin — COMMENTED** [review]
looks pretty good! I have a few qs to make sure I understand the code. You could ignore all the "nit:" comments/suggestions/questions.

File: test/jit/test_dtype_analysis.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+        input_shapes = [
+            ((1, 1, 1), (1, 1)),  # Different Dim, non-zerodim
+            ((), (1, 2)),  # One zerodim
+            ((1, 2), ()),  # Other zerodim
+            ((), ()),  # both zerodim
+        ]
+
+        input_dtypes = [
+            (float32, float32),  # Simple Case
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply test/jit/test_dtype_analysis.py:136]: Yeah, I consider it covered by the combination of (float, float), and (int32, int64). 
(again, this test case is already taking multiple seconds, I am really worried about expanding it further).

File: test/jit/test_dtype_analysis.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            # Scalar vs Scalar is automatically inferred.
+        ]
+
+        input_dtypes = [
+            (float32, float32),  # Simple Case
+            (int32, int64),  # Size Promotion (compliated case for 0dim tensors)
+            (int32, float32),  # type Promotion
+        ]
+
+        with set_default_dtype(float32):
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply test/jit/test_dtype_analysis.py:161]: I don't see what we gain by that.

File: torch/csrc/jit/passes/dtype_analysis.cpp [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+  // for efficiency
+  bool processAtenOps(Node* n) {
+    GRAPH_DEBUG("processAtenOps");
+    GRAPH_DEBUG("case = ", n->kind(), " ", *n);
+    bool changed = tryApplyDtypeMetaTensor(n);
+
+    switch (n->kind()) {
+      case aten::append:
+        // auto elementType =
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/passes/dtype_analysis.cpp:256]: Will actually just return False, as we will just not process that for now.

File: torch/csrc/jit/passes/dtype_analysis.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    GRAPH_DEBUG("Argument missing Dtype");
+    return false;
+  }
+  auto rank = tensor_type->symbolic_sizes().rank();
+  return rank.has_value();
+}
+
+static bool canBeInferredWithMetaTensor(Node* n) {
+  // Not a guarantee that the metatensor will not error out
+  // Do not have a allowlist for now and let things error out in execution.
```
[the remark this answers is not in the document]
Comment (Gamrix) [reply torch/csrc/jit/passes/dtype_analysis.cpp:69]: Also, the custom rules are basically also intended to be a blocklist.

File: test/jit/test_dtype_analysis.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            ((1, 1, 1), (1, 1)),  # Different Dim, non-zerodim
+            ((), (1, 2)),  # One zerodim
+            ((1, 2), ()),  # Other zerodim
+            ((), ()),  # both zerodim
+        ]
+
+        input_dtypes = [
+            (float32, float32),  # Simple Case
+            (int32, int64),  # Size Promotion (compliated case for 0dim tensors)
+            (float32, int32),  # type Promotion
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/jit/test_dtype_analysis.py:139]: makes sense.

File: test/jit/test_dtype_analysis.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/jit/test_dtype_analysis.py:162]: fair enough

File: torch/csrc/jit/passes/dtype_analysis.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+  if (!tensor_type->scalarType().has_value()) {
+    value->setType(tensor_type->withScalarType(scalarType));
+    return true;
+  }
+  if (tensor_type->scalarType().value() != scalarType) {
+    // Have to overwrite the dtype if it has been set to a diffferent value
+    // in a previous run of dtype analysis (eg in testing with cached graphs)
+
+    // Due to SSA, there shouldn't be a reason why this overwrite is not valid.
+    value->setType(tensor_type->withScalarType(scalarType));
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply torch/csrc/jit/passes/dtype_analysis.cpp:123]: looks good

File: test/jit/test_dtype_analysis.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        graph_out = list(graph.outputs())
+        assert(len(graph_out) == 1)
+        return graph_out[0].type().dtype()
+
+    def prop_dtype_on_graph(self, graph, example_inputs):
+        graph_inputs = list(graph.inputs())
+
+        # We need to clear shape information because torch.jit.script
+        # will return a cached graph if the function is scripted twice.
+        torch._C._jit_pass_erase_shape_information(graph)
```
Comment (Krovatkin) [inline test/jit/test_dtype_analysis.py:42]: awesome!!!!

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175961).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175961).

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32175961).

**Pull request closed without merging**